Update obsproc build for RDHPCS spack-stack and Rocky OS

KateFriedman-NOAA opened the pull request:
This PR updates the `dev/gfsv17` branch of obsproc to build with spack-stack installs, which were made post-Rocky8[9] transitions on the R&Ds. This PR also includes an updated spack-stack install on Jet in `/contrib` after moving away from the `lfs4` filesystem.

Similar to updates made for prepobs in its `dev/gfsv17` branch, the build script is updated and the build/run version files are broken out for each supported platform. Since spack-stack (in it's current incarnation) is not yet on WCOSS2, the WCOSS2 build does not source `obsproc_common.lua` and instead has all of the library module versions loaded in the same modulefile. This should be updated ahead of cutting the obsproc tag for supporting GFSv17.

Resolves #88

KateFriedman-NOAA committed:
Update obsproc build for spack-stack and Rocky OS

- Add build/run version files for each supported platform
- Add version and modulefiles for new platform (Hercules)
- Update build to use new version files
- Update .gitignore

Refs #88

KateFriedman-NOAA committed:
Merge remote-tracking branch 'upstream/dev/gfsv17' into dev/gfsv17

* upstream/dev/gfsv17:
  Add module purge for building

KateFriedman-NOAA committed:
Ignore build[run].ver

- Add temporarily ver files to .gitignore
- Remove build.ver and run.ver

Refs #88

KateFriedman-NOAA committed:
Updates for WCOSS2

Retain functionality on WCOSS2

Refs #88

KateFriedman-NOAA committed:
Update R&D modulefiles to use MODULEPATH variables

Refs #88

KateFriedman-NOAA committed:
Remove BUILD_4NCO from build

Refs #88

KateFriedman-NOAA committed:
Update spack-stack path on Jet

Use new /contrib location after move off /lfs4

Refs #88

aerorahul commented on modulefiles/obsproc_hera.lua:
```diff
@@ -2,12 +2,15 @@ help([[
-load("hpc/1.1.0")
-load("hpc-intel/18.0.5.274")
-load("hpc-impi/2018.0.4")
+stack_intel_ver=os.getenv("stack_intel_ver") or "2021.5.0"
+stack_impi_ver=os.getenv("stack_impi_ver") or "2021.5.1"
+cmake_ver=os.getenv("cmake_ver") or "3.23.1"
```
I suggest we remove the defaults here; i.e. the `or` part for spack installed stuff.  They are either defined in `build.spack.ver` or in `build.platform.ver`. That way, these modulefiles will likely only see updates to the `MODULEPATH`, which I think we should also define in `build.platform.ver` to something like `spack_modpath=....`. Take away all the gory details and clump them in `build.platform.ver` and `run.platform.ver`.
These module files then are free of any specifics.  Thoughts?

KateFriedman-NOAA replied:
Agreed. Removed defaults @ dcd0a43.

aerorahul commented on modulefiles/obsproc_hercules.lua:
```diff
@@ -0,0 +1,22 @@
+load("obsproc_common")
+
+-- On Hercules, MKL needs to be loaded separately
+load("intel-oneapi-mkl/2022.2.1")
```
this should be `intel_mkl_ver`, no?

aerorahul replied:
also, this is different from the one specified in `run.hercules.ver`

KateFriedman-NOAA replied:
Corrected @ 2276cc4, thanks for catching!

aerorahul commented on modulefiles/obsproc_hera.lua:
```diff
@@ -2,12 +2,15 @@ help([[
 ]])
 
-load("cmake/3.20.1")
+prepend_path("MODULEPATH", pathJoin("/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-" .. (os.getenv("spack_stack_ver") or "None"), "/envs/" .. (os.getenv("spack_env") or "None"), "-dev-rocky8/install/modulefiles/Core"))
```
```suggestion
prepend_path("MODULEPATH", os.getenv("spack_stack_mod_path"))
```
`export spack_stack_mod_path=" .... /install/modulefiles/Core"` in `build.hera.ver` and `run.hera.ver`.  
Yes, that will be repetition, but that is already true with all the versions that are in `run.spack.ver` and `build.spack.ver`.  Ideally, we should have a single `spack.ver` and source it in `build.platform.ver` and `run.platform.ver`.  Setting a version is ok even if it is not used in `build` or `run`.  E.g. having `prod_util_ver` in build step or having `cmake_ver` in run, is harmless.
For the future.  @DavidHuber-NOAA thoughts?

DavidHuber-NOAA replied:
Sounds like a great idea.  Opened noaa-emc/global-workflow#2844.

KateFriedman-NOAA replied:
Suggested modifications made @ e097418

KateFriedman-NOAA committed:
Remove defaults from modulefiles

Refs #88

KateFriedman-NOAA committed:
Correct intel-oneapi-mkl module load for Hercules

- Update to include version variable

Refs #88

KateFriedman-NOAA committed:
Merge spack.ver and introduce spack_stack_mod_path

- Combine build.spack.ver and run.spack.ver into spack.ver
- Set spack_stack_mod_path in ver files and use variable
for MODULEPATH in modulefiles

Refs #88

aerorahul reviewed (approved):
Thanks for making the suggested changes.  These look clean.

KateFriedman-NOAA commented:
Successfully retested build on WCOSS2, Hera, Jet, Orion, and Hercules.

DavidHuber-NOAA commented on modulefiles/obsproc_common.lua:
```diff
@@ -2,17 +2,19 @@ help([[
-sigio_ver=os.getenv("sigio_ver") or "default"
-nemsio_ver=os.getenv("nemsio_ver") or "default"
-bufr_ver=os.getenv("bufr_ver") or "default"
+hdf5_ver=os.getenv("hdf5_ver")
```
If `hdf5_ver` is not in the environment, then `hdf5_ver` will be set to an empty string in this module file.  This will result in the default module being loaded below, which may or may not be the one you want to use.  If this is not desired, then the only way I have found for this to always raise an error is to provide a non-existent version if `hdf5_ver` is not in the environment, e.g.

```lua
hdf5_ver=os.getenv("hdf5_ver") or "None"
```
This will cause a failure when attempting to load the `hdf5/None` module.

KateFriedman-NOAA replied:
Thanks for this info @DavidHuber-NOAA ! Question, would this also be a concern for the other module versions or just specifically `hdf5`?

DavidHuber-NOAA replied:
It would be a concern for all versions, I just used `hdf5` as an example.

KateFriedman-NOAA replied:
Per your suggestion, I have added `or "None"` to the modulefiles when obtaining the version variables from the environment. Committed @ b6cbdf1

KateFriedman-NOAA committed:
Set "None" for module versions to avoid default

To avoid loading a default module version when the
module version is not set, put "None" as the default
so as to raise an error at build/run time if the
version is not set.

Refs #88

DavidHuber-NOAA commented on versions/spack.ver:
```diff
@@ -0,0 +1,22 @@
+
+export cmake_ver=3.23.1
+
+export jasper_ver=2.0.32
+export libpng_ver=1.6.37
+export zlib_ver=1.2.13
+
+export hdf5_ver=1.14.0
+export netcdf_c_ver=4.9.2
+export netcdf_fortran_ver=4.6.1
+
+export bacio_ver=2.4.1
+export nemsio_ver=2.5.4
+export sp_ver=2.5.0
+export sigio_ver=2.3.2
+export w3emc_ver=2.10.0
+export bufr_ver=11.7.0
+
+export grib_util_ver=1.3.0
```
Some of these libraries are not loaded.  There's likely no harm in including them, but suggest paring this down to only the loaded modules:

```suggestion
export hdf5_ver=1.14.0
export netcdf_c_ver=4.9.2
export netcdf_fortran_ver=4.6.1

export bacio_ver=2.4.1
export nemsio_ver=2.5.4
export sp_ver=2.5.0
export sigio_ver=2.3.2
export w3emc_ver=2.10.0
export bufr_ver=11.7.0

```

ilianagenkova replied:
It looks like spack.ver is not used for wcoss2, right?

KateFriedman-NOAA replied:
Correct. It is not used _yet_. Until we have the newer spack-stack installed on WCOSS2 we won't use `spack.ver` on WCOSS2.

ilianagenkova commented on .gitignore:
```diff
@@ -15,10 +15,12 @@ fort.*
+build*/
+exec*/
+install*/
+versions/build.ver
```
Why are /versions/*.ver to be be ignored?

KateFriedman-NOAA replied:
I have created machine-specific `.ver` files. At build time the specific `.ver` files for the machine get copied to `build.ver` and `run.ver` (to satisfy the NCO requirement). Since these are copies of existing files and not something we want to commit back, we should ignore them in the git status.

ilianagenkova commented on modulefiles/obsproc_common.lua:
```diff
@@ -2,17 +2,19 @@ help([[
 
 load(pathJoin("hdf5", hdf5_ver))
-load(pathJoin("netcdf", netcdf_ver))
+load(pathJoin("netcdf-c", netcdf_c_ver))
```
Is "netcdf-c" the same as "netcdf-C"?

KateFriedman-NOAA replied:
@AlexanderRichert-NOAA can you answer Iliana's question?

@ilianagenkova here is the `module show` for the `netcdf-c/4.9.2` (on Jet):
```
[Kate.Friedman@fe6 v1.1.3]$ module show netcdf-c/4.9.2
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
   /misc/contrib/spack-stack/spack-stack-1.6.0/envs/gsi-addon-intel/install/modulefiles/intel-oneapi-mpi/2021.5.1/intel/2021.5.0/netcdf-c/4.9.2.lua:
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
whatis("Name : netcdf-c")
whatis("Version : 4.9.2")
whatis("Target : core2")
whatis("Short description : NetCDF (network Common Data Form) is a set of software libraries and machine-independent data formats that support the creation, access, and sharing of array-oriented scientific data. This is the C distribution.")
help([[Name   : netcdf-c]])
help([[Version: 4.9.2]])
help([[Target : core2]])
help()
help([[NetCDF (network Common Data Form) is a set of software libraries and
machine-independent data formats that support the creation, access, and
sharing of array-oriented scientific data. This is the C distribution.]])
depends_on("c-blosc/1.21.5")
depends_on("curl/8.4.0")
depends_on("hdf5/1.14.0")
depends_on("zlib/1.2.13")
depends_on("zstd/1.5.2")
prepend_path("PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/bin")
prepend_path("MANPATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/share/man")
prepend_path("LD_LIBRARY_PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/lib")
prepend_path("DYLD_LIBRARY_PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/lib")
prepend_path("CPATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/include")
prepend_path("PKG_CONFIG_PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/lib/pkgconfig")
prepend_path("CMAKE_PREFIX_PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/.")
prepend_path("PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/bin")
prepend_path("MANPATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/share/man")
prepend_path("PKG_CONFIG_PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/lib/pkgconfig")
prepend_path("CMAKE_PREFIX_PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/.")
append_path("HDF5_PLUGIN_PATH","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24/plugins")
setenv("netcdf_c_ROOT","/misc/contrib/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/netcdf-c-4.9.2-wxvro24")
append_path("MANPATH","")
```

DavidHuber-NOAA replied:
@KateFriedman-NOAA @ilianagenkova netcdf-c is the C-language netCDF library.  Spack-stack installs netcdf-c and netcdf-fortran separately to allow for increased flexibility.

This is different than the `netcdf-C` module on WCOSS2, which contains both the C- and Fortran-language libraries.  Moreover, the `-A`, `-B`, and `-C` libraries on WCOSS2 are tied to three HPC-stack co-installations of a handful of packages (HDF5, netCDF, ESMF, and pnetcdf) and should be used together (e.g. hdf5-A should be loaded with netcdf-A and esmf-A).

ilianagenkova replied:
Glad I asked! Thanks for the explanation, @DavidHuber-NOAA

KateFriedman-NOAA replied:
Good to know, thanks @DavidHuber-NOAA !

ilianagenkova commented on modulefiles/obsproc_common.lua:
```diff
@@ -2,17 +2,19 @@ help([[
 load(pathJoin("hdf5", hdf5_ver))
-load(pathJoin("netcdf", netcdf_ver))
+load(pathJoin("netcdf-c", netcdf_c_ver))
+load(pathJoin("netcdf-fortran", netcdf_fortran_ver))
```
This module doesn't seem to be available on wcoss. Won't that be a problem?

KateFriedman-NOAA replied:
I updated things so that `obsproc_common.lua` isn't loaded on WCOSS2 (temporary measure until we have the newer spack-stack installed there and can begin loading the newer module names like `netcdf-c` and `netcdf-fortran`). All of the modules for WCOSS2 are loading within it's primary modulefile for now.

ilianagenkova replied:
Thanks!

ilianagenkova commented on modulefiles/obsproc_wcoss2.lua:
```diff
@@ -2,11 +2,20 @@ help([[
-cmake_ver=os.getenv("cmake_ver")
-craype_ver=os.getenv("craype_ver")
-cray_mpich_ver=os.getenv("cray_mpich_ver")
+PrgEnv_intel_ver=os.getenv("PrgEnv_intel_ver") or "None"
```
Has this been tested on WCOSS2, i.e. the "None" option?

KateFriedman-NOAA replied:
Yes, as a test I commented out the `PrgEnv_intel_ver=8.2.0` line in `versions/build.wcoss2.ver`:
```
diff --git a/versions/build.wcoss2.ver b/versions/build.wcoss2.ver
index e3859b2..24b2aab 100644
--- a/versions/build.wcoss2.ver
+++ b/versions/build.wcoss2.ver
@@ -7,7 +7,7 @@ export sigio_ver=2.3.2
 export nemsio_ver=2.5.2
 export bufr_ver=11.7.0
 
-export PrgEnv_intel_ver=8.2.0 #8.1.0
+#export PrgEnv_intel_ver=8.2.0 #8.1.0
 export intel_ver=19.1.3.304
 export cmake_ver=3.20.2
 export craype_ver=2.7.13 #2.7.8
```
...and then got an error during build because it couldn't load that module and that module was needed:
```
+ cmake -DCMAKE_INSTALL_PREFIX=/lfs/h2/emc/global/noscrub/kate.friedman/git/obsproc/dev-gfsv17 -DCMAKE_INSTALL_BINDIR=exec -DCMAKE_BUILD_TYPE=Release ..
-- The Fortran compiler identification is unknown
Error: A PrgEnv-* modulefile must be loaded.
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - failed
```

ilianagenkova commented on ush/build.sh:
```diff
@@ -7,57 +7,46 @@ readonly pkg_root=$(cd "$(dirname "$(readlink -f -n "${BASH_SOURCE[0]}" )" )/.."
+  cp "build.${target}.ver" "build.ver"
+  cp "run.${target}.ver"   "run.ver"
+  cd "${pkg_root}" || exit 1
+  export HOMEobsproc="${pkg_root}"
```
Is HOMEobsproc going to point at current package's location?

KateFriedman-NOAA replied:
Yes, `HOMEobsproc` will be the clone location for a moment during build before being unset.

ilianagenkova reviewed (approved):
Could you answer my questions before your merge these changes?

ilianagenkova commented:
Thanks, @KateFriedman-NOAA and @DavidHuber-NOAA!

KateFriedman-NOAA committed:
Remove unneeded versions in spack.ver

Remove jasper, libpng, zlib, and grib_util versions

Refs #88

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

DavidHuber-NOAA reviewed (approved):
Looks good, thanks!
